specs/CameraCaptureUI/CameraCaptureUI.md

```diff
@@ -0,0 +1,387 @@
+
+| Name | Value | Description |
+|-|-|-|
+| HighestAvailable | 0 | The user can select any resolution. |
+| LowDefinition | 1 | The user can select resolutions up to low definition resolutions.|
+| StandardDefinition | 2 | The user can select resolutions up to standard definition resolutions.|
+| HighDefinition | 3 | The user can select resolutions up to high definition resolutions. |
+
+
+
+
+
+
+# API Details
+
+> Note: all of this new WinAppSDK API is the same as the existing UWP
+[CameraCaptureUI](https://docs.microsoft.com/uwp/api/Windows.Media.Capture.CameraCaptureUI)
+API, only changing the constructor to take a
+[WindowId](https://docs.microsoft.com/windows/windows-app-sdk/api/winrt/Microsoft.UI.WindowId).
+
+ 
+```c++ (but really MIDL3)
+namespace Microsoft.Windows.Media.Capture
+{
+    [contractversion(1)]
+    apicontract CameraCaptureUIContract {};
+ 
+    [contract(CameraCaptureUIContract, 1)]
+    typedef enum CameraCaptureUIMode
+    {
+        PhotoOrVideo = 0,
+        Photo,
+        Video,
+    } CameraCaptureUIMode;
+ 
+    [contract(CameraCaptureUIContract, 1)]
+    typedef enum CameraCaptureUIPhotoFormat
+    {
+        Jpeg = 0,
+        Png,
+        JpegXR,
+    } CameraCaptureUIPhotoFormat;
+ 
+    [contract(CameraCaptureUIContract, 1)]
+    typedef enum CameraCaptureUIVideoFormat
+    {
+        Mp4 = 0,
+        Wmv,
+    } CameraCaptureUIVideoFormat;
+ 
+    [contract(CameraCaptureUIContract, 1)]
+    typedef enum CameraCaptureUIMaxVideoResolution
+    {
+        HighestAvailable = 0,
+        LowDefinition,
+        StandardDefinition,
+        HighDefinition,
+    } CameraCaptureUIMaxVideoResolution;
+ 
+    [contract(CameraCaptureUIContract, 1)]
+    typedef enum CameraCaptureUIMaxPhotoResolution
+    {
+        HighestAvailable = 0,
+        VerySmallQvga,
+        SmallVga,
+        MediumXga,
+        Large3M,
+        VeryLarge5M,
+    } CameraCaptureUIMaxPhotoResolution;
+ 
+    [contract(CameraCaptureUIContract, 1)]
+    [default_interface]
+    runtimeclass CameraCaptureUIPhotoCaptureSettings
+    {
+        CameraCaptureUIPhotoCaptureSettings();
+ 
+        Boolean AllowCropping;
+        Windows.Foundation.Size CroppedAspectRatio;
+        Windows.Foundation.Size CroppedSizeInPixels;
+        Microsoft.Windows.Media.Capture.CameraCaptureUIPhotoFormat Format;
+        Microsoft.Windows.Media.Capture.CameraCaptureUIMaxPhotoResolution MaxResolution;
+    }
+ 
+    [contract(CameraCaptureUIContract, 1)]
+    [default_interface]
+    runtimeclass CameraCaptureUIVideoCaptureSettings
+    {
+        CameraCaptureUIVideoCaptureSettings();
+ 
+        Boolean AllowTrimming;
+        Microsoft.Windows.Media.Capture.CameraCaptureUIVideoFormat Format;
+        Single MaxDurationInSeconds;
+        Microsoft.Windows.Media.Capture.CameraCaptureUIMaxVideoResolution MaxResolution;
+    }
+ 
+    [contract(CameraCaptureUIContract, 1)]
+    [default_interface]
+    [threading(sta),
```

Comment:
So this threading model means that this API must be called from a UI thread?  What happens if an app has multiple islands with multiple UI threads?  Is that something they need to worry about or is the implementation threadsafe?

Reply:
Note that multi-UI-threaded is an existing scenario with the system API.

Reply:
I observed in the OS CameraCaptureUI Api, particularly in the CaptureFileAsync function, there is a locking mechanism implemented to only allow one operation at a time. 

I figure this is something I would need to implement for my API as well.

Reply:
FWIW, https://www.windowslatest.com/2024/08/21/windows-11-24h2-will-finally-allow-multiple-apps-to-use-your-camera-at-the-same-time/ (since, winui3 apps _will be_ multi-UI-threaded.)

Reply:
The locking mechanism in the UWP CameraCapture API operates at the object level, allowing only one capture operation per instance. I've added a task to track this internally and will implement a similar implementation for the WinAppSDK version of CameraCapture.

@mominshaikhdevs, the update that allows multiple apps to use the camera simultaneously seems to be outside the scope of this API. It doesn't appear that any changes are necessary at the CameraCaptureUI level. The primary function of CaptureFileAsync is to open the camera and return a storage file object. Thanks for bringing this to my attention.

Reply:
The API is designed to be used from a UI thread due to its STA threading model requirement. In a multi-UI-threaded environment, each thread should individually follow the STA model. If the API is called from different UI threads, it should work as long as each call is made from an STA-thread. I have tested the API in a WinUI3 packaged blank app multi-window scenario and it works as expected.